### PR TITLE
Add Template tag tests

### LINKED ISSUE #44: Add templatetag tests
Verify expected operation of the various template tags.

### PR DESCRIPTION
**Pull request by kasulani**
* Add tests for template tags
* Fixes #44

### CODE REVIEW

File: pinax/ratings/templatetags/pinax_ratings_tags.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def overall_rating(object, category=""):
 def rating_post_url(user, obj):
     ct = ContentType.objects.get_for_model(obj)
     post_url = reverse(
-        "pinax_ratings_rate",
+        "pinax_ratings:rate",
```
Comment (grahamu) [inline pinax/ratings/templatetags/pinax_ratings_tags.py]: nice catch!

Comment (kasulani) [reply pinax/ratings/templatetags/pinax_ratings_tags.py]: Thanks @grahamu

File: pinax/ratings/tests/test_tags.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    """
+
+    def setUp(self):
+        self.test_user = self.make_user(username="test_user")
+        self.another_user = self.make_user(username="another_user")
+        self.none_car_object = self.make_user(username="a_user")
+        self.benz = Car.objects.create(name="Mercedes c200")
+        self.post_a_rating(self.test_user, 2)
+
+    def post_a_rating(self, user, rating):
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            "{% load pinax_ratings_tags %}"
+            "{% user_rating user object category as the_rating %}"
+        )
+        context = Context({
+            "user": self.test_user,
+            "object": self.benz,
+            "category": "handling"
+        })
+        template.render(context)
+        self.assertEqual(context["the_rating"], 2)
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    """
+    Tests for the template tags
+    """
+
+    def setUp(self):
+        self.test_user = self.make_user(username="test_user")
+        self.another_user = self.make_user(username="another_user")
+        self.none_car_object = self.make_user(username="a_user")
+        self.benz = Car.objects.create(name="Mercedes c200")
+        self.post_a_rating(self.test_user, 2)
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                    "category": "handling"
+                },
+            )
+
+    def test_user_rating_tag_with_category(self):
+        """
+        Ensure the template tag user_rating renders a rating
+        as posted by the user on a Car object they are rating
+        for a specified category in the context
+        """
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            "{% load pinax_ratings_tags %}"
+            "{% user_rating user object category as the_rating %}"
+        )
+        context = Context({
+            "user": self.test_user,
+            "object": self.benz,
+            "category": ""
+        })
+        template.render(context)
+        self.assertEqual(context["the_rating"], 2)
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        template = Template(
+            "{% load pinax_ratings_tags %}"
+            "{% overall_rating object category as the_overall_rating %}"
+        )
+        context = Context({
+            "object": self.benz,
+            "category": "handling"
+        })
+        template.render(context)
+        self.assertEqual(context["the_overall_rating"], 2.5)
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        template = Template(
+            "{% load pinax_ratings_tags %}"
+            "{% overall_rating object category as the_overall_rating %}"
+        )
+        context = Context({
+            "object": self.benz,
+            "category": ""
+        })
+        template.render(context)
+        self.assertEqual(context["the_overall_rating"], 2.0)
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+        template = Template(
+            "{% load pinax_ratings_tags %}"
+            "{% rating_count object as count %}"
+        )
+        context = Context({
+            "object": self.benz
+        })
+        template.render(context)
+        self.assertEqual(context["count"], 1)
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

**Comment by grahamu** [comment]
@kasulani I'm somewhat confused... you posted "Done @grahamu as requested" on all my requests three days ago, but there is no new code since ten days ago. Did you forget to push your commits?

**Comment by grahamu** [comment]
Aha, now I realize your PR is not complete @kasulani. I'll wait to hear from you.

File: pinax/ratings/tests/test_tags.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+            "{% load pinax_ratings_tags %}"
+            "{% user_rating_js user object %}"
+        )
+        context = Context({
+            "user": self.test_user,
+            "object": self.benz,
+            "category": "handling"
+        })
+        rendered_template = template.render(context)
+        self.assertInHTML(
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+            "{% user_rating_url user object as url %}"
+        )
+        context = Context({
+            "user": self.test_user,
+            "object": self.benz
+        })
+        template.render(context)
+        self.assertEqual(
+            context["url"],
+            "/{}/{}/rate/".format(
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: Done @grahamu as requested

File: pinax/ratings/tests/test_tags.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+        template = Template(
+            "{% load pinax_ratings_tags %}"
+            "{% ratings object as a_rating %}"
+        )
+        context = Context({
+            "object": self.benz
+        })
+        template.render(context)
+        self.assertEqual(context["a_rating"], [])
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_tags.py]: @grahamu please see slack channel for my response to this review comment

**Comment by kasulani** [comment]
@grahamu I have pushed some commits to fix the issues you pointed out in the review

**Comment by grahamu** [comment]
@kasulani your tests are greatly improved! You've correctly addressed all but one of my requests.

The only remaining issue is `test_ratings_tag`, where you are testing the wrong assertion.

Upon review, my [earlier comment](https://github.com/pinax/pinax-ratings/pull/47#discussion_r167414424) is partially incorrect:

> This test passes, but I do not believe it should. Write tests to prove an assertion, not to pass. In this case the test docstring is wrong and the "ratings" ttag works improperly.

So the first two sentences are correct, but the third, asserting "ratings" ttag code is wrong, is wrong.
I'm now thinking that "ratings" ttag is correct, it just isn't properly tested.

> First correct the docstring by examining the "ratings" ttag code and deciding what it should return in this situation. Then write assertions proving the expected value is returned.

This is still true, and this is where you missed my direction. Examine the "ratings" ttag code and see what is getting returned. The way I see it, this ttag is returning a QuerySet of Ratings, and the test should verify the queryset.

> I believe properly constructed assertions will fail and the code should be corrected.

Pretty sure I'm wrong here... pretty sure "ratings" ttag code is OK.

> Hint: verifying queryset equality can be tricky. I typically verify same number of items in each QuerySet, then convert both QuerySets to sets and verify set equality.

Yep. This test should not render a template. Instead examine and verify the return value from "ratings" ttag, similar to what you've done with `test_user_rating_js_tag`. I typically use `assertSetEqual` for queryset equality (when duplicates are not expected) because it ignores object order:

```
ct = ContentType.objects.get_for_model(self.benz)
output = rating(<args>)
expected = Rating.objects.filter(content_type=ct, object_id=self.benz.pk)
self.assertSetEqual(set(output), set(expected))
```

So one more update and I think this will be ready for merge. Thanks so much for your efforts @kasulani!

**Comment by grahamu** [comment]
BTW @kasulani you are making great progress, and significantly improving test coverage for pinax-ratings, as seen in this coverage report:
screen shot 2018-03-14 at 2 25 39 pm

**Pull request merged**